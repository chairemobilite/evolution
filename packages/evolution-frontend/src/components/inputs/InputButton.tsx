/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { ChoiceType, InputButtonType } from '../../services/widgets';
import * as surveyHelper from '../../utils/helpers';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { FrontendUser } from 'chaire-lib-frontend/lib/services/auth/user';
import { WithTranslation, withTranslation } from 'react-i18next';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export interface InputButtonProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    id: string;
    onValueChange: (e: any) => void;
    value: string;
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: FrontendUser;
    inputRef?: React.LegacyRef<HTMLTextAreaElement>;
    size?: 'small' | 'medium' | 'large';
    widgetConfig: InputButtonType;
    closeQuestionModal?: (path: string) => void;
    questionModalPath?: string;
}

export interface InputButtonOptionProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    id: string;
    choice: ChoiceType;
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: FrontendUser;
    onButtonClick: (e: React.MouseEvent, value: string) => void;
}

const InputButtonOption = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputButtonOptionProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation
) => {
    if (
        props.choice.hidden === true ||
        !surveyHelper.parseBoolean(props.choice.conditional, props.interview, props.path, props.user)
    ) {
        return null;
    }
    return (
        <div key={`${props.id}__key`} className={'survey-question__input-button-input-container'}>
            <button
                type="button"
                className={`survey-section__button button ${props.choice.color || 'green'} ${
                    props.choice.size || 'large'
                }`}
                onClick={(e) => props.onButtonClick(e, props.choice.value)}
                id={props.id}
            >
                {props.choice.icon && <FontAwesomeIcon icon={props.choice.icon} className="faIconLeft" />}
                {surveyHelper.parseString(
                    props.choice.label[props.i18n.language] || props.choice.label,
                    props.interview,
                    props.path
                )}
            </button>
        </div>
    );
};

const InputButtonOptionT = withTranslation()(InputButtonOption);

/** Display a list of choices as buttons */

const InputButton = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputButtonProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation
) => {
    const widgetChoices = props.widgetConfig.choices;
    const choices =
        typeof widgetChoices === 'function' ? widgetChoices(props.interview, props.path, props.user) : widgetChoices;

    // FIXME: original code had an action parameter, that was part of the choice object. It isn't used anywhere though and the saveCallback and valueChange should be enough to cover the case. If not, revisit this.
    const onButtonClick = (e: React.MouseEvent, value: string) => {
        e.preventDefault();
        props.onValueChange({ target: { value } });

        if (props.widgetConfig.isModal === true && props.closeQuestionModal) {
            props.closeQuestionModal(props.questionModalPath || '');
        }
    };

    const choiceInputs = choices.map((choice, index) => (
        <InputButtonOptionT
            key={`${props.id}_btn_${index}`}
            id={`${props.id}__input-button__${choice.value}`}
            choice={choice}
            path={props.path}
            interview={props.interview}
            user={props.user}
            onButtonClick={onButtonClick}
        />
    ));

    return (
        <div
            className={`survey-question__input-button-group-container${
                props.widgetConfig.sameLine === false ? ' no-wrap' : ''
            }${props.widgetConfig.align === 'center' ? ' align-center' : ''}`}
        >
            {choiceInputs}
        </div>
    );
};

export default withTranslation()(InputButton);
