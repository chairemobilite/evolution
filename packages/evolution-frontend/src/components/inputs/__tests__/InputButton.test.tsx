/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';

import { interviewAttributes } from './interviewData.test';
import InputButton from '../InputButton';
import i18next from 'i18next';

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

describe('InputButton with normal options', () => {
    const conditionalFct = jest.fn().mockReturnValue(true);
    const translationFct = jest.fn().mockReturnValue('Translated string');
    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' },
            hidden: false,
            color: 'green'
        },
        {
            value: 'val2',
            label: 'Unilingual label',
            iconPath: 'img/test.png',
            color: 'red',
            size: 'small' as const
        },
        {
            value: 'hiddenVal',
            label: { en: 'english hidden', fr: 'cachée' },
            hidden: true
        },
        {
            value: 'conditionalVal',
            label: { en: 'english conditional', fr: 'conditionnelle' },
            conditional: conditionalFct
        },
        {
            value: 'val3',
            label: translationFct,
            hidden: false,
            color: 'green'
        }
    ];

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'button' as const,
        choices,
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        },
        isModal: true
    };

    test('Render InputButton with normal option type', () => {
        
        const wrapper = TestRenderer.create(
            <InputButton
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value='value'
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
        expect(conditionalFct).toHaveBeenCalledTimes(1);
        expect(conditionalFct).toHaveBeenCalledWith(interviewAttributes, 'foo.test', userAttributes);
        expect(translationFct).toHaveBeenCalledWith(i18next.t, interviewAttributes, 'foo.test', userAttributes);
        
    });

    test('Test Button click event', () => {
        const mockOnValueChange = jest.fn();
        const { queryByText } = render(
            <InputButton
                id={'test'}
                onValueChange={mockOnValueChange}
                widgetConfig={widgetConfig}
                value='value'
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        // Html was added, so the complete text is not there, it is actually composed of many texts
        const buttonOption1 = queryByText(choices[0].label['en']);
        expect(buttonOption1).toBeTruthy();

        const buttonOption2 = queryByText(choices[1].label as string);
        expect(buttonOption2).toBeTruthy();

        const buttonOption3 = queryByText(choices[3].label['en']);
        expect(buttonOption3).toBeTruthy();
    
        // Click on the close button
        fireEvent.click(buttonOption1 as any);
        expect(mockOnValueChange).toHaveBeenCalledTimes(1);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: choices[0].value }});

        // Click on the close button
        fireEvent.click(buttonOption2 as any);
        expect(mockOnValueChange).toHaveBeenCalledTimes(2);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: choices[1].value }});

        // Click on the close button
        fireEvent.click(buttonOption3 as any);
        expect(mockOnValueChange).toHaveBeenCalledTimes(3);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: choices[3].value }});
    });

    test('Test Button click event with close modal', () => {
        const mockOnValueChange = jest.fn();
        const mockCloseModal = jest.fn();
        const { queryByText } = render(
            <InputButton
                id={'test'}
                onValueChange={mockOnValueChange}
                widgetConfig={widgetConfig}
                value='value'
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
                closeQuestionModal={mockCloseModal}
                questionModalPath={'modalPath'}
            />
        );
        // Html was added, so the complete text is not there, it is actually composed of many texts
        const buttonOption1 = queryByText(choices[0].label['en']);
        expect(buttonOption1).toBeTruthy();
    
        // Click on the close button
        fireEvent.click(buttonOption1 as any);
        expect(mockOnValueChange).toHaveBeenCalledTimes(1);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: choices[0].value }});
        expect(mockCloseModal).toHaveBeenCalledTimes(1);
        expect(mockCloseModal).toHaveBeenCalledWith('modalPath');
    })
})

test('Render InputButton with choice function', () => {

    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' },
            hidden: false,
            color: 'green'
        },
        {
            value: 'val2',
            label: 'Unilingual label',
            iconPath: 'img/test.png',
            color: 'red',
            size: 'small' as const
        },
        {
            value: 'hiddenVal',
            label: { en: 'english hidden', fr: 'cachée' },
            hidden: true
        },
        {
            value: 'conditionalVal',
            label: { en: 'english conditional', fr: 'conditionnelle' },
            conditional: () => false
        }
    ];
    const choiceFct = jest.fn().mockReturnValue(choices);
    
    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'button' as const,
        choices: choiceFct,
        align: 'center' as const,
        sameLine: false,
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };
    
    const wrapper = TestRenderer.create(
        <InputButton
            id={'test'}
            onValueChange={() => { /* nothing to do */}}
            widgetConfig={widgetConfig}
            value='value'
            inputRef={React.createRef()}
            size='medium'
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
        />
    );
    expect(wrapper).toMatchSnapshot();
    expect(choiceFct).toHaveBeenCalledTimes(1);
    expect(choiceFct).toHaveBeenCalledWith(interviewAttributes, 'foo.test', userAttributes);
    
});
