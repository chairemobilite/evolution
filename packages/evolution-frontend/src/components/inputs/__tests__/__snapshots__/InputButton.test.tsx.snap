// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputButton with normal options Render InputButton with normal option type 1`] = `
<div
  className="survey-question__input-button-group-container"
>
  <div
    className="survey-question__input-button-input-container"
  >
    <button
      className="survey-section__button button green large"
      id="test__input-button__val1"
      onClick={[Function]}
      type="button"
    >
      english value
    </button>
  </div>
  <div
    className="survey-question__input-button-input-container"
  >
    <button
      className="survey-section__button button red small"
      id="test__input-button__val2"
      onClick={[Function]}
      type="button"
    >
      Unilingual label
    </button>
  </div>
  <div
    className="survey-question__input-button-input-container"
  >
    <button
      className="survey-section__button button green large"
      id="test__input-button__conditionalVal"
      onClick={[Function]}
      type="button"
    >
      english conditional
    </button>
  </div>
</div>
`;

exports[`Render InputButton with choice function 1`] = `
<div
  className="survey-question__input-button-group-container no-wrap align-center"
>
  <div
    className="survey-question__input-button-input-container"
  >
    <button
      className="survey-section__button button green large"
      id="test__input-button__val1"
      onClick={[Function]}
      type="button"
    >
      english value
    </button>
  </div>
  <div
    className="survey-question__input-button-input-container"
  >
    <button
      className="survey-section__button button red small"
      id="test__input-button__val2"
      onClick={[Function]}
      type="button"
    >
      Unilingual label
    </button>
  </div>
</div>
`;
